angular-frontend/src/app/authorization/authorization.service.ts

```diff
@@ -0,0 +1,229 @@
+import {Injectable} from '@angular/core';
+import {HttpClient, HttpHeaders, HttpParams, HttpRequest} from '@angular/common/http';
+import {Observable} from 'rxjs';
+import {Router} from "@angular/router";
+import jwt_decode from 'jwt-decode';
+import {AccessTokenResponse, DecodedToken} from './AuthorizationTypes';
+import {randomBytes} from 'crypto';
+import * as cryptojs from "crypto";
+import base64url from 'base64url';
+import {Utils} from "../utils";
+import {AppConfigService} from "../app-config.service";
+
+@Injectable({
+  providedIn: 'root'
+})
+export class AuthorizationService {
+  private readonly authorizationEndpoint = '/oauth2/authorize';
+  private readonly codeChallengeMethod = 'S256';
+  private readonly responseType = 'code';
+  private readonly authorizationCode = 'authorization_code';
+  private readonly accessTokenRef = 'access_token';
+  protected authority: string | undefined
+  protected clientId: string | undefined
+  protected scope: string[] | undefined
+
+  constructor(
+    private http: HttpClient,
+    private router: Router,
+    private utils: Utils,
+    private appConfig: AppConfigService
+  ) {}
+
+  ngOnInit(): void {
+    this.authority = this.appConfig.getAuthority()
+    this.clientId = this.appConfig.getClientId()
+    this.scope = this.ensureContainsScopeOpenId(this.appConfig.getScope())
+  }
+
+  async startAuthentication(): Promise<any> {
+    this.ensureConfigLoaded()
+    const scope = this.scope?.join(' ') ?? 'openid';
+    const clientId = this.clientId ?? 'default-frontend-client-id-change-me';
+    const redirectUri = this.utils.removeQueryParamsFromUrl(this.utils.getCurrentUrl());
+
+    // Construct authorization request URL
+    const params = new URLSearchParams({
+      response_type: this.responseType,
+      client_id: clientId,
+      redirect_uri: redirectUri,
+      scope: scope,
+      state: this.getState(),
+      nonce: this.getNonce(),
+      code_challenge_method: this.codeChallengeMethod,
+      code_challenge: this.getCodeChallenge(),
+    });
+
+    const authUrl = this.authority + this.authorizationEndpoint + '?' + params.toString();
```

Comment:
**[issue]**: this type of URL construction may be susceptible to attack (e.g. XSS). I would use something a wrapping function `encodeURI` to construct a safe authUrl